feat: blog: Using AST to build rules in ESLint (#324)
eslint/eslint.org: src/content/drafts/sourcecode-to-tokens-AST.md

---
layout: post
title: How a parser translates text into tokens and then into an abstract syntax tree

> teaser: "You will learn how the source code in any proramming language is converted to token and then parsed to generate an AST. The AST can then be traversed to write and build custom rules in ESLint."

Reviewer: ```suggestion
teaser: "You will learn how the source code in any programming language is converted to a token and then parsed to generate an AST. The AST can then be traversed to write and build custom rules in ESLint."
```

Author: Since the source code is converted to multiple tokens, a token may not be applicable.
I will make the change suitably.

Reviewer: sure.

image: 
authors:

>   - Gayathri Krishnaswamy

Reviewer 2: Needs to be your GitHub username.

```suggestion
  - gayathri-krishnaswamy
```

Author: Fixed

categories:

>   - Blog post

Reviewer 2: We formally put blog posts into specific categories. This is a tutorial, so:

```suggestion
  - Tutorials
```

Author: Fixed.

tags:
  - Lexical analysis
  - Syntax analysis
  - Abstract Syntax Tree
  - AST
  - Tokens
  - ESLint rules
---
##  Introduction
[ESLint](https://eslint.org/docs/latest/) is an open-source pluggable JavaScript linter that uses the Abstract Syntax Tree (AST) of your source code to identify and report patterns found in ECMAScript/JavaScript code. The tool also provides an option to write and configure your own rules to enforce a particular coding style that is relevant only to your project.

In this article, first, we'll see on a high-level how your source code is converted into multiple tokens and is then parsed to produce an AST that represents the syntactic structure of your source code. Next, we'll understand how ESLint leverages the capability of AST and helps you build your own custom rules.

Before we jump into details, let's understand a couple of key terminologies:

> - Compiler - A [compiler](https://en.wikipedia.org/wiki/Compiler) is a special program that converts the programming language’s source code into machine-readable code that targets specific operating systems and computer architectures. 

Reviewer 2: If we are going to introduce compilers as part of this blog, then I think we need to make it clear a compiler uses a parser.

Author: I have modified the Introduction and removed references to compilers. Please let me know if this is ok.

A parser is a program that takes input in the form of sequential source program instructions and breaks it into smaller pieces (tokens) to generate statements and expressions. The parser then represents the statements and expressions in the form of a parse tree or an Abstract Syntax Tree (AST) that depicts the syntactic structure of your source code.

In this tutorial, first, you'll learn on a high-level how your source code is broken down into multiple tokens and is then converted to build an AST. Next, you'll learn how [ESLint](https://eslint.org/docs/latest/) leverages the capability of AST and helps you build your own custom rules.

The main steps involved in how the parsers process the source code are:

Step 1: Lexical analysis

Step 2: Syntax analysis

Reviewer 2: The audience is someone who uses ESLint and wants to understand more about how it works. Generally, I think what you have here is a good introduction. One thing that is worth calling out is that parsers do not all necessary create an AST. An AST is created only when someone wants to be able to inspect the overall structure of a program. Parsers written for other purposes, such as immediate conversion into another format, may not create an AST at all.

Author: Okay. I have modified the Introduction as follows:

A parser is a program that takes input in the form of sequential source program instructions and breaks it into smaller pieces (tokens) to generate statements and expressions. The parser then represents the statements and expressions in the form of a parse tree or an Abstract Syntax Tree (AST) that depicts the syntactic structure of your source code.

:information_source: Not all parsers create an AST. An AST is required only to inspect the overall structure of the program. Parsers written for other purposes, such as immediate conversion into another format, may not create an AST at all.

In this tutorial, first, you'll learn on a high-level how your source code is broken down into multiple tokens and is then converted to an AST. Next, you'll learn how [ESLint](https://eslint.org/docs/latest/) leverages the capability of AST and helps you build your own custom rules. This tutorial is intended for those who use ESLint and wants to understand more about how the tool works. It assumes that you have basic familiarity with:

- Computer programming
- Javascript coding
- ESLint tool

By the end of this tutorial, you’ll be able to:

-  Understand how the source code is converted to AST.
- Generate AST for your source program using AST explorer.
- Understand how to traverse the AST nodes to write custom rules in ESLint.

> - Parser - A [parser](https://en.wikipedia.org/wiki/Parsing) is a program that is part of the compiler that generates a very abstract representation of your source code, often a tree or a graph.

Reviewer 2: Can we get more into details here? A parser is actually made up of multiple steps. It starts out with tokenization and then uses that information to generate statements and expressions. Last, parsers *might* represent that information in an abstract syntax tree.

Author: Please refer https://github.com/eslint/eslint.org/pull/324#discussion_r953403330

> Now, let's look into the main steps involved in how compilers process the source code:

Reviewer 2: Again, because ESLint isn't a compiler, I don't think compilers should be the focus of the post. Parsers should be the focus.

Author: Please refer https://github.com/eslint/eslint.org/pull/324#discussion_r953403330

Step 1: Lexical analysis

Step 2: Syntax analysis

Step 3: Code generation

> 

Reviewer: ```suggestion
```

Author: Resolved.

:information_source:**Tip**: If you ever get lost in the AST, you can hover over different sections of your source code on the right and the tool then highlights the relevant node in the AST on the left.

The AST represents the relationship between the tokens generated from the source code as a tree comprising different nodes. Every node has at least a type specifying what it is representing. For example, a node type could be a `Literal` representing an actual value or a `CallExpression` that represents a function call. The `Literal` node might only contain a value whereas the `CallExpression` node can contain additional information such as what is being called (`callee`), or what `arguments` are being passed in.

:memo: It is important to remember that the format of AST changes depending on the type of programming language and the type of tool used for parsing. 

###  Linting using AST
You can traverse the different nodes of the AST to gain insights or perform actions. One of the most common use cases that extensively use AST traversal is

> [ linting](https://en.wikipedia.org/wiki/Lint_(software)). ESLint uses [Espree](https://github.com/eslint/espree)  parser to create an AST from JavaScript code and the AST is defined by the  common standard [ESTree](https://github.com/estree/estree) specification. To understand the different types of nodes represented in the AST refer to [ES5 AST grammar documentation](https://github.com/estree/estree/blob/master/es5.md). You can refer to ESLint[ documentation](https://eslint.org/docs/latest/user-guide/configuring/rules#disabling-rules) to learn more about their built-in rules and how you can build custom rules, plugins, and formatters.

Reviewer: ```suggestion
[ linting](https://en.wikipedia.org/wiki/Lint_(software)). ESLint uses [Espree](https://github.com/eslint/espree)  parser to create an AST from JavaScript code and the AST is defined by the common standard [ESTree](https://github.com/estree/estree) specification. To understand the different types of nodes represented in the AST refer to [ES5 AST grammar documentation](https://github.com/estree/estree/blob/master/es5.md). You can refer to ESLint[ documentation](https://eslint.org/docs/latest/user-guide/configuring/rules#disabling-rules) to learn more about their built-in rules and how you can build custom rules, plugins, and formatters.
```

> Here's an[ example in the AST explorer](https://astexplorer.net/#/gist/052d25ec2db5e45442e474ffc8dd0f43/f2146de2d27649296a810fc4e88293b649782649) that establishes a rule that you cannot have an empty catch block. We then traverse the AST and write a rule in ESLint to check if there are empty catch blocks in the source code.

Reviewer: ```suggestion
Here's an[ example in the AST explorer](https://astexplorer.net/#/gist/052d25ec2db5e45442e474ffc8dd0f43/f2146de2d27649296a810fc4e88293b649782649) that establishes a rule that you cannot have an empty catch block. We then traverse the AST and write a rule in ESLint to check if there are empty catch blocks in the source code.

```

```
"use strict";
module.exports = {
  meta: {
    type: 'problem', // `problem`, `suggestion`, or `layout`
    docs: {
      description: "This rule catches an empty catch block",
       },
    fixable: null, // Or `code` or `whitespace`
    schema: [], // Add a schema if the rule has options
    messages: {
      emptyCatch: 'Empty catch block is not allowed.',
  },
  },
  create(context) {
    return {
      CatchClause(node)
        {
          if (node.body.body.length === 0) {
            context.report({node: node.body, messageId: 'emptyCatch'});
          }
        }
    };
  },
};

```

In the above code snippet, we check if the length of the body of the `CatchClause` node is 0. If yes,  ESLint’s [report](https://eslint.org/docs/latest/developer-guide/working-with-rules#contextreport) method is used to raise a complaint.

## Step 3: Code generation

> Need inputs from engineers.

Reviewer 2: I think we can skip code generation. It's not important for ESLint.

Author: Fixed.

---
layout: post
title: How a parser translates text into tokens and then into an abstract syntax tree

> teaser: "You will learn how the source code in any proramming language is converted to multiple tokens and is then parsed to generate an Abstract Syntax Tree. You can then then traverse the AST to write and build custom rules in ESLint."

Author: Fixed

- Computer programming
- Javascript coding
- ESLint tool
  
By the end of this tutorial, you’ll be able to:
- Understand how the source code is converted to AST.
- Generate AST for your source program using AST explorer.
- Understand how to traverse the AST nodes to write custom rules in ESLint.

## Step 1: Lexical analysis
[Lexical analysis](https://en.wikipedia.org/wiki/Lexical_analysis) commonly known as tokenization is the process of converting your source code into a sequence of tokens. A token can be any valid entity in a programming language such as an identifier, keyword, operator, function call, string, or separator. Tokens do not understand how things fit together instead focus on different components of a file. Consider it analogous to breaking a sentence into different words and identifying the words such as nouns, verbs, numbers, adjectives, and punctuation. At this point, you don’t understand what’s part of a sentence or how these words fit together to make a sentence. The below picture depicts how a simple `if statement` is converted into multiple tokens.

> ![If statement converted to tokens](/src/assets/images/blog/2022/lexical-analysis.png "source code to tokens")

Reviewer 3: https://github.com/eslint/eslint.org/pull/324/files#r954747406
```suggestion
![If statement converted to tokens](/assets/images/blog/2022/lexical-analysis.png "source code to tokens")
```

Author: Changed the path. Can you please check in preview mode?

## Step 2: Syntax analysis
In this step, the list of tokens is converted into an AST - tree-like data structure that represents the actual structure of the code.

> ![AST](/src/assets/images/blog/2022/AST-visual.png "AST")

Reviewer 3: Please start the image URL with `/assets`. Images are right now not being displayed in the preview

```suggestion
![AST](/assets/images/blog/2022/AST-visual.png "AST")
```

Reviewer 3: Oh... but make sure we change the path during publishing :)

Author: Changed the path. Can you please check in preview mode?

Reviewer 3: https://deploy-preview-324--new-eslint.netlify.app/content/drafts/sourcecode-to-tokens-ast/ It works fine now

The AST visual shown above is generated using [JointJS](https://resources.jointjs.com/demos/javascript-ast).

 You can consider this step analogous to finding the relationship between nouns, verbs, and adjectives in a sentence and identifying how they conform to the language’s grammar.  Continuing with the same example, in the previous step, there were three individual tokens for x, >, and 18 but after syntax analysis, it can be inferred that x>18 is a condition for an `if statement`. 

> ![A list of tokens converted to AST](/src/assets/images/blog/2022/syntax-analysis.png "tokens to AST")

Reviewer 3: https://github.com/eslint/eslint.org/pull/324/files#r954747406
```suggestion
![A list of tokens converted to AST](/assets/images/blog/2022/syntax-analysis.png "tokens to AST")
```

Author: Changed the path. Can you please check in preview mode?

In the above example,[ AST explorer](https://astexplorer.net/#/gist/052d25ec2db5e45442e474ffc8dd0f43/latest) is used for visualizing the AST for a simple JavaScript code snippet. AST explorer is a freely available online tool that can be used to generate AST for different programming languages.

> :information_source:**Tip**: If you ever get lost in the AST, you can hover over different sections of your source code on the right and the tool then highlights the relevant node in the AST on the left.

Reviewer 3: `:information_source:` Just wanted to confirm if we were trying to transform this into different icon or something else? But it renders as such in preview.

Author: :information_source: is a markdown emoji. It will be rendered as ℹ️  I thought instead of writing **Note**  I could use this emoji. It does not render in the VSCode preview though.

Reviewer 3: I did not use VSCode preview. I used the netlify build preview and it didn't show up. 
https://deploy-preview-324--new-eslint.netlify.app/content/drafts/sourcecode-to-tokens-ast/

Author: Okay! Can I change it to just Note?

Reviewer 3: I see in other parts of the documentation we have used `**Note:**` may be we can use the same

Author: I have changed the emoji to **Note:**
